src/test/socket/shutdown/test_shutdown.rs

```diff
@@ -0,0 +1,974 @@
+                    |s| format!("{} <domain={},type={},flag={}>", s, domain, sock_type, flag);
+
+                tests.extend(vec![test_utils::ShadowTest::new(
+                    &append_args("test_not_connected"),
+                    move || test_not_connected(domain, sock_type, flag),
+                    vec![TestEnv::Libc, TestEnv::Shadow],
+                )]);
+            }
+        }
+    }
+
+    for &domain in domains.iter() {
+        for &flag in flags.iter() {
+            // add details to the test names to avoid duplicates
+            let append_args = |s| format!("{} <domain={},flag={}>", s, domain, flag);
+
+            tests.extend(vec![test_utils::ShadowTest::new(
+                &append_args("test_tcp_fin_correctness"),
+                move || test_tcp_fin_correctness(domain, flag),
+                vec![TestEnv::Libc, TestEnv::Shadow],
+            )])
+        }
+    }
+
+    tests
+}
+
+/// Test shutdown() using an argument that cannot be a fd.
+fn test_invalid_fd() -> Result<(), String> {
+    let args = ShutdownArguments {
+        fd: -1,
+        how: libc::SHUT_RDWR,
+    };
+
+    check_shutdown_call(&args, &[libc::EBADF])
+}
+
+/// Test shutdown() using an argument that could be a fd, but is not.
+fn test_non_existent_fd() -> Result<(), String> {
+    let args = ShutdownArguments {
+        fd: 8934,
+        how: libc::SHUT_RDWR,
+    };
+
+    check_shutdown_call(&args, &[libc::EBADF])
+}
+
+/// Test shutdown() using a valid fd that is not a socket.
+fn test_non_socket_fd() -> Result<(), String> {
+    let args = ShutdownArguments {
+        fd: 0, // assume the fd 0 is already open and is not a socket
+        how: libc::SHUT_RDWR,
+    };
+
+    check_shutdown_call(&args, &[libc::ENOTSOCK])
+}
+
+/// Test shutdown() using a valid socket, but invalid `how` argument.
+fn test_invalid_how() -> Result<(), String> {
+    let fd = unsafe { libc::socket(libc::AF_INET, libc::SOCK_STREAM, 0) };
+    assert!(fd >= 0);
+
+    let args = ShutdownArguments { fd, how: 88 };
+
+    test_utils::run_and_close_fds(&[fd], || check_shutdown_call(&args, &[libc::EINVAL]))
+}
+
+/// Test shutdown() using a non-connected socket.
+fn test_not_connected(
+    domain: libc::c_int,
+    sock_type: libc::c_int,
+    flag: libc::c_int,
+) -> Result<(), String> {
+    let fd = unsafe { libc::socket(domain, sock_type | flag, 0) };
+    assert!(fd >= 0);
+
+    let args = ShutdownArguments {
+        fd,
+        how: libc::SHUT_RDWR,
+    };
+
+    test_utils::run_and_close_fds(&[fd], || check_shutdown_call(&args, &[libc::ENOTCONN]))
+}
+
+/// Generate a pair of connected TCP sockets.
+fn setup_stream_sockets(domain: libc::c_int, flag: libc::c_int) -> (libc::c_int, libc::c_int) {
+    let fd_client = unsafe { libc::socket(domain, libc::SOCK_STREAM | flag, 0) };
+    let fd_listener = unsafe { libc::socket(domain, libc::SOCK_STREAM | flag, 0) };
+    assert!(fd_client >= 0);
+    assert!(fd_listener >= 0);
+
+    {
```

Comment:
Optional throughout, but I still slightly prefer keeping things a bit more contact by inlining the function call rather than introducing a local variable and limiting its lifetime with a new scope. e.g.

```
assert_eq!(unsafe { libc::listen(fd_listener, 10) }, 0);
```

Reply:
I'm still not convinced that this is more readable, especially with larger function calls and if you have assert messages, but I converted the ones that will fit on a single line.

Reply:
There's definitely a bit of a tradeoff here. The actual assertion might be a little easier to understand with the temp variable and new scope, but adding so many more lines has a cost of making the code more sparse, making it difficult to look at as much at once and see the overall structure, etc.

I like how you have it now; I agree that it becomes more worth it to break it up with longer / more-complex calls and assertions.